pipeline-runner/R/gem2s-3-run_emptydrops.R

```diff
@@ -1,64 +1,47 @@
+#'
+#' @inheritParams download_cellranger
+#' @param prev_out 'output' slot from call to \code{load_cellranger}
+#'
+#' @return \code{prev_out} with added slot 'edrops' containing result of call to
+#'   \code{\link[DropletUtils]{emptyDrops}} for each sample.
+#'
+#' @export
+#'
+run_emptydrops <- function(input, pipeline_config, prev_out) {
+  message("Testing if droplets are empty...")
 
-  scdata_list <- scdata_list[samples]
+  # destructure previous output
+  counts_list <- prev_out$counts_list
+  samples <- names(counts_list)
 
-  message("calculating probability of barcodes being background noise...")
-  for (sample_name in names(scdata_list)) {
-    compute_emptydrops(scdata_list[[sample_name]], sample_name)
+  edrops <- list()
+  for (sample in samples) {
+    message("\nSample --> ", sample)
+    edrops[[sample]] <- compute_emptydrops(counts_list[[sample]])
   }
 
-  message("Step 3 completed.")
+  prev_out$edrops <- edrops
+  res <- list(
+    data = list(),
+    output = prev_out)
 
-  return(list())
+  message("\nRunning of emptydrops step complete.")
+  return(res)
 }
 
 
-# compute_emptydrops function
-#' @description Save the result of emptyDrops per sample.
-#' @param scdata Raw sparse matrix with the counts for one sample.
-#' @param sample_name Name of the sample that we are preparing.
-#'
-#' @return
-compute_emptydrops <- function(scdata, sample_name) {
-  message("Sample --> ", sample_name, "...")
+#' @param counts dgCMatrix with counts for one sample.
+compute_emptydrops <- function(counts) {
+  # check if filtered
+  nempty <- sum(Matrix::colSums(counts) < gem2s$max.empty.counts)
 
-  # [HARDCODE]
-  threshold_emptydrops <- 100
-  # threshold_emptydrops <- 2000 # use this to simulate unfiltered data
-  # automatically checking if data contains enough (50) barcodes that
-  # are confidently considered to be empty to serve as training set
-  nr_barcodes_ambient <- sum(colSums(scdata) < threshold_emptydrops)
-  range(colSums(scdata))
-  if (nr_barcodes_ambient < 50) {
-    flag_filtered <- TRUE
-    message("Found not enough [", nr_barcodes_ambient, "] ambient barcodes,
-          so data is considered to be pre-filtered")
-    message("Smallest nr of UMI for any barcode: ", min(colSums(scdata)))
-    message("EmptyDrops filter will not be applied")
+  if (nempty < gem2s$max.empty.drops) {
+    message("Detected sample as filtered --> Skipping emptyDrops.")
+    out <- NULL
   } else {
-    flag_filtered <- FALSE
-    # if non-filtered, run emptyDrops and save results in file
-    message("Found enough [", nr_barcodes_ambient, "] ambient barcodes,
-          so data is considered to be not filtered")
-    emptydrops_out <- DropletUtils::emptyDrops(scdata, lower = threshold_emptydrops)
-
-    file_path <- paste("/output/pre-emptydrops-", sample_name, ".rds", sep = "")
-    saveRDS(emptydrops_out, file = file_path, compress = FALSE)
+    out <- DropletUtils::emptyDrops(counts)
   }
+
+  return(out)
```

Comment:
small thing, but I recommend small renamings to make it more obvious what this function does:
- rename `out`: give it a name to show that it is the empty droplets for a sample. For example, "sample_edrops".
- rename the function parameter `counts` to signify that these are per sample. For example, "sample_counts".
- rename function `compute_emptydrops` to `compute_sample_emptydrops` or even `compute_sample_edrops` so that it is consistent with the choice of `edrops` for empty droplets in all other places.

Reply:
Good suggestion - done :)